main/buildgen/src/mill/main/buildgen/BuildGenUtil.scala

```diff
@@ -0,0 +1,363 @@
+package mill.main.buildgen
+
+import mill.main.buildgen.BuildObject.Companions
+import mill.main.client.CodeGenConstants.{
+  buildFileExtensions,
+  nestedBuildFileNames,
+  rootBuildFileNames,
+  rootModuleAlias
+}
+import mill.main.client.OutFiles
+import mill.runner.FileImportGraph.backtickWrap
+
+@mill.api.internal
+object BuildGenUtil {
+
+  type BuildNode = Node[BuildObject]
+  type BuildPackage = String
+  type BuildTree = Tree[BuildNode]
+
+  type Gav = (String, String, String)
+  type InterpIvy = String
+
+  type BomIvyDeps = IterableOnce[String] // interpolated or named
+  type IvyDeps = IterableOnce[String] // inmergeterpolated or named
+  type ModuleDeps = IterableOnce[BuildPackage]
+
+  def buildFile(dirs: Seq[String]): os.SubPath = {
+    val name = if (dirs.isEmpty) rootBuildFileNames.head else nestedBuildFileNames.head
+    os.sub / dirs / name
+  }
+
+  def buildFiles(workspace: os.Path): geny.Generator[os.Path] =
+    os.walk.stream(workspace, skip = (workspace / OutFiles.out).equals)
+      .filter(file => buildFileExtensions.contains(file.ext))
+
+  def buildPackage(dirs: Seq[String]): BuildPackage =
+    (rootModuleAlias +: dirs).iterator.map(backtickWrap).mkString(".")
+
+  def buildPackages[Module, Key](input: Tree[Node[Module]])(key: Module => Key)
+      : Map[Key, BuildPackage] =
+    input.nodes.fold(Map.newBuilder[Key, BuildPackage])((z, node) =>
+      z += ((key(node.module), buildPackage(node.dirs)))
+    ).result()
+
+  def buildSource(node: BuildNode): os.Source = {
+    val pkg = buildPackage(node.dirs)
+    val BuildObject(imports, companions, supertypes, inner, outer) = node.module
+    val importStatements = imports.iterator.map("import " + _).mkString(linebreak)
+    val companionTypedefs = companions.iterator.map {
+      case (_, vals) if vals.isEmpty => ""
+      case (name, vals) =>
+        val members =
+          vals.iterator.map { case (k, v) => s"val $k = $v" }.mkString(linebreak)
+
+        s"""object $name {
+           |
+           |$members
+           |}""".stripMargin
+    }.mkString(linebreak2)
+
+    s"""package $pkg
+       |
+       |$importStatements
+       |
+       |$companionTypedefs
+       |
+       |object `package` ${mkExtends(supertypes)} {
+       |
+       |$inner
+       |}
+       |
+       |$outer
+       |""".stripMargin
+  }
+
+  def compact(tree: BuildTree): BuildTree = {
+    println("compacting Mill build tree")
```

Comment:
Maybe leave a code comment here explaining what "compacting" means and why we want to do it?